packages/muelu/src/Graph/MatrixTransformation/MueLu_CutDrop.hpp

```diff
@@ -0,0 +1,515 @@
+  }
+
+  template <class local_matrix_type2, class DistanceFunctorType2, class diag_view_type2>
+  struct Comparator {
+   private:
+    using scalar_type        = typename local_matrix_type2::value_type;
+    using local_ordinal_type = typename local_matrix_type2::ordinal_type;
+    using memory_space       = typename local_matrix_type2::memory_space;
+    using results_view       = Kokkos::View<DecisionType*, memory_space>;
+
+    using ATS           = Kokkos::ArithTraits<scalar_type>;
+    using magnitudeType = typename ATS::magnitudeType;
+
+    const local_matrix_type2 A;
+    const diag_view_type2 diag;
+    const DistanceFunctorType2* dist2;
+    const local_ordinal_type rlid;
+    const local_ordinal_type offset;
+    const results_view results;
+
+    const scalar_type one = ATS::one();
+
+   public:
+    KOKKOS_INLINE_FUNCTION
+    Comparator(const local_matrix_type2& A_, const diag_view_type2& diag_, const DistanceFunctorType2* dist2_, local_ordinal_type rlid_, const results_view& results_)
+      : A(A_)
+      , diag(diag_)
+      , dist2(dist2_)
+      , rlid(rlid_)
+      , offset(A_.graph.row_map(rlid_))
+      , results(results_) {}
+
+    KOKKOS_INLINE_FUNCTION
+    magnitudeType get_value(size_t x) const {
+      auto clid = A.graph.entries(offset + x);
+      scalar_type val;
+      if (rlid != clid) {
+        val = one / dist2->distance2(rlid, clid);
+      } else {
+        val = diag(rlid);
+      }
+      auto aiiajj = ATS::magnitude(diag(rlid)) * ATS::magnitude(diag(clid));  // |a_ii|*|a_jj|
+      auto aij2   = ATS::magnitude(val) * ATS::magnitude(val);                // |a_ij|^2
+      return (aij2 / aiiajj);
+    }
+
+    KOKKOS_INLINE_FUNCTION
+    bool operator()(size_t x, size_t y) const {
+      if (results(offset + x) != UNDECIDED) {
+        if (results(offset + y) != UNDECIDED) {
+          // does not matter
+          return (x < y);
+        } else {
+          // sort undecided to the right
+          return true;
+        }
+      } else {
+        if (results(offset + y) != UNDECIDED) {
+          // sort undecided to the right
+          return false;
+        } else {
+          return get_value(x) > get_value(y);
+        }
+      }
+    }
+  };
+
+  using comparator_type = Comparator<local_matrix_type, DistanceFunctorType, diag_view_type>;
+
+  KOKKOS_INLINE_FUNCTION
+  comparator_type getComparator(local_ordinal_type rlid) const {
+    return comparator_type(A, diag, &dist2, rlid, results);
+  }
+};
+
+template <class view_type, class comparator_type>
+KOKKOS_INLINE_FUNCTION void serialHeapSort(view_type& v, comparator_type comparator) {
+  auto N       = v.extent(0);
+  size_t start = N / 2;
+  size_t end   = N;
+  while (end > 1) {
+    if (start > 0)
+      start = start - 1;
+    else {
+      end       = end - 1;
+      auto temp = v(0);
+      v(0)      = v(end);
+      v(end)    = temp;
+    }
+    size_t root = start;
+    while (2 * root + 1 < end) {
```

Comment:
@jhux2

Reply:
My concern was related to shell sort, not heap sort.  Question:  is a stable sort needed here?

Reply:
I don't think so? If two entries give the same value for sorting then either both should be kept or dropped.